src/pipecat/processors/canonical_metrics_processor.py

```diff
@@ -0,0 +1,202 @@
+        await super().process_frame(frame, direction)
+        if self.end_of_call:
+            return
+
+        if (isinstance(frame, EndFrame) or isinstance(frame, CancelFrame)):
+            self.end_of_call = True
+            if self.has_audio():
+                os.makedirs(self.output_dir, exist_ok=True)
+                filename = self.get_output_filename()
+                with BytesIO() as buffer:
+                    with wave.open(buffer, 'wb') as wf:
+                        wf.setnchannels(self.num_channels)
+                        wf.setsampwidth(self.sample_rate // 8000)
+                        wf.setframerate(self.sample_rate)
+                        wf.writeframes(self.audio_buffer)
+                    wave_data = buffer.getvalue()
+
+                async with aiofiles.open(filename, 'wb') as file:
+                    await file.write(wave_data)
+
+                try:
+                    await self.multipart_upload(filename)
+                    await aiofiles.os.remove(filename)
+                except FileNotFoundError:
+                    pass
+                except Exception as e:
+                    raise e
+                self.audio_buffer = bytearray()
+
+    def get_output_filename(self):
+        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+        return f"{self.output_dir}/{timestamp}-{uuid.uuid4().hex}.wav"
+
+    def canonical_api_url(self):
+        return os.environ.get("CANONICAL_API_URL", "https://voiceapp.canonical.chat/api/v1")
+
+    def request_headers(self):
+        return {
+            "Content-Type": "application/json",
+            "X-Canonical-Api-Key": os.environ.get("CANONICAL_API_KEY")
+        }
+
+    async def multipart_upload(self, file_path: str):
+        upload_request, upload_response = await self.request_upload(file_path)
+        parts = await self.upload_parts(file_path, upload_request, upload_response)
+        await self.upload_complete(parts, upload_request, upload_response)
+
+    async def request_upload(self, file_path: str) -> Tuple[Dict, Dict]:
+        filename = os.path.basename(file_path)
+        filename = f"{str(uuid.uuid4())}-{filename}"
+        filesize = os.path.getsize(file_path)
+        numparts = int((filesize + self.partsize - 1) / self.partsize)
+
+        params = {
+            'filename': filename,
+            'parts': numparts,
+            'assistant': self.assistant,
+            'assistantSpeaksFirst': self.assistant_speaks_first
+        }
+        print(f"Requesting presigned URLs for {numparts} parts")
+        async with aiohttp.ClientSession() as session:
+            async with session.post(
+                f"{self.canonical_api_url()}/recording/uploadRequest",
+                headers=self.request_headers(),
+                json=params
+            ) as response:
+                if not response.ok:
+                    raise Exception(f"Failed to get presigned URLs: {await response.text()}")
```

Comment:
Here we could use `logger.error` and return so the bot doesn't stop working.

Reply:
Mmm... I actually think we capture the exception upstream, so maybe this is OK. I'll double-check

Reply:
Done.